### PR TITLE
Ix: Improve IgnoreElements() performance

### PR DESCRIPTION
**Pull request by akarnokd**
This PR reduces the overhead in Ix `IgnoreElements` by not doing a full `foreach` but only `MoveNext` on the upstream's `IEnumerator`.

```
BenchmarkDotNet=v0.10.14, OS=Windows 7 SP1 (6.1.7601.0)
Intel Core i7-4770K CPU 3.50GHz (Haswell), 1 CPU, 8 logical and 4 physical cores

Frequency=3418037 Hz, Resolution=292.5656 ns, Timer=TSC
.NET Core SDK=2.1.301
  [Host]     : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT
```

#### Before

 Method |       N |            Mean |          Error |         StdDev |  Gen 0 | Allocated |
------- |--------:|----------------:|---------------:|---------------:|-------:|----------:|
 Ignore |       1 |        50.70 ns |      0.4997 ns |      0.4674 ns | 0.0362 |     152 B |
 Ignore |      10 |        93.22 ns |      0.2499 ns |      0.2215 ns | 0.0361 |     152 B |
 Ignore |     100 |       516.74 ns |      1.9667 ns |      1.8397 ns | 0.0353 |     152 B |
 Ignore |    1000 |     4,685.37 ns |     16.8983 ns |     15.8067 ns | 0.0305 |     152 B |
 Ignore |   10000 |    46,111.36 ns |     97.3665 ns |     91.0766 ns |      - |     152 B |
 Ignore |  100000 |   460,584.23 ns |  1,447.6994 ns |  1,354.1788 ns |      - |     152 B |
 Ignore | 1000000 | 4,601,262.98 ns | 23,059.4917 ns | 21,569.8607 ns |      - |     152 B |
 
#### After
 
 Method |       N |            Mean |          Error |         StdDev |  Gen 0 | Allocated |
------- |--------:|----------------:|---------------:|---------------:|-------:|----------:|
 Ignore |       1 |        50.08 ns |      0.4549 ns |      0.4033 ns | 0.0362 |     152 B |
 Ignore |      10 |        77.11 ns |      0.4092 ns |      0.3628 ns | 0.0361 |     152 B |
 Ignore |     100 |       352.52 ns |      1.5889 ns |      1.4862 ns | 0.0358 |     152 B |
 Ignore |    1000 |     3,058.96 ns |     18.0046 ns |     15.0346 ns | 0.0343 |     152 B |
 Ignore |   10000 |    29,961.48 ns |    101.0439 ns |     89.5728 ns | 0.0305 |     152 B |
 Ignore |  100000 |   297,692.90 ns |  1,586.6047 ns |  1,484.1109 ns |      - |     152 B |
 Ignore | 1000000 | 2,982,819.68 ns | 11,058.0834 ns | 10,343.7370 ns |      - |     152 B |

### CODE REVIEW

File: Ix.NET/Source/System.Interactive/IgnoreElements.cs [hunk 1]
```diff
@@ -26,12 +26,11 @@ public static IEnumerable<TSource> IgnoreElements<TSource>(this IEnumerable<TSou
 
         private static IEnumerable<TSource> IgnoreElements_<TSource>(this IEnumerable<TSource> source)
         {
-            foreach (var item in source)
-            {
-                ;
-            }
+            var enumerator = source.GetEnumerator();
```
Comment (onovotny) [inline Ix.NET/Source/System.Interactive/IgnoreElements.cs]: I believe we still have to dispose this, so a try/finally block?

Comment (akarnokd) [reply Ix.NET/Source/System.Interactive/IgnoreElements.cs]: Indeed. I'll update the PR and remeasure the benefits.

Comment (akarnokd) [reply Ix.NET/Source/System.Interactive/IgnoreElements.cs]: Updated the measurements. Still looks promising.

File: Ix.NET/Source/System.Interactive/IgnoreElements.cs [hunk 2]
```diff
@@ -26,12 +26,13 @@ public static IEnumerable<TSource> IgnoreElements<TSource>(this IEnumerable<TSou
 
         private static IEnumerable<TSource> IgnoreElements_<TSource>(this IEnumerable<TSource> source)
         {
-            foreach (var item in source)
```
Comment (onovotny) [inline Ix.NET/Source/System.Interactive/IgnoreElements.cs]: One other thing to test -- what is the impact on `List<T>` or array's? IIRC, `foreach` has optimizations beyond this.

Comment (akarnokd) [reply Ix.NET/Source/System.Interactive/IgnoreElements.cs]: Sure will try it a bit later. The question is, should we expect side-effects when enumerating a List? As for arrays, we can simply throw them away as there should be no side-effect when traversing and ignoring them. Also there is the conceptional question about what's the purpose of ignoring known lists and arrays?

Comment (onovotny) [reply Ix.NET/Source/System.Interactive/IgnoreElements.cs]: I don't expect side effects, but the code gen/implementation of foreach vs using the enumerator directly could be slightly different

**Comment by akarnokd** [comment]
Added benchmarks for list and array sources.

#### Before

Method |       N |            Mean |          Error |         StdDev |  Gen 0 | Allocated |
------------ |--------:|----------------:|---------------:|---------------:|-------:|----------:|
Ignore |       1 |        47.36 ns |      0.7308 ns |      0.5706 ns | 0.0362 |     152 B |
IgnoreList |       1 |        45.95 ns |      0.3561 ns |      0.3157 ns | 0.0362 |     152 B |
IgnoreArray |       1 |        39.04 ns |      0.2439 ns |      0.2163 ns | 0.0343 |     144 B |
Ignore |      10 |        88.58 ns |      0.2004 ns |      0.1875 ns | 0.0361 |     152 B |
IgnoreList |      10 |       106.83 ns |      0.3054 ns |      0.2857 ns | 0.0362 |     152 B |
IgnoreArray |      10 |        76.90 ns |      0.3295 ns |      0.2921 ns | 0.0343 |     144 B |
Ignore |     100 |       515.36 ns |      1.3841 ns |      1.2947 ns | 0.0353 |     152 B |
IgnoreList |     100 |       624.86 ns |      2.6833 ns |      2.5099 ns | 0.0362 |     152 B |
IgnoreArray |     100 |       488.34 ns |      2.3555 ns |      2.2033 ns | 0.0343 |     144 B |
Ignore |    1000 |     4,727.26 ns |      7.3105 ns |      6.4806 ns | 0.0305 |     152 B |
IgnoreList |    1000 |     5,779.42 ns |     15.8669 ns |     14.8419 ns | 0.0305 |     152 B |
IgnoreArray |    1000 |     4,452.37 ns |     13.2589 ns |     12.4024 ns | 0.0305 |     144 B |
Ignore |   10000 |    46,737.52 ns |    112.0152 ns |    104.7791 ns |      - |     152 B |
IgnoreList |   10000 |    57,239.52 ns |    134.5450 ns |    119.2706 ns |      - |     152 B |
IgnoreArray |   10000 |    43,721.34 ns |     95.9446 ns |     85.0524 ns |      - |     144 B |
Ignore |  100000 |   466,920.08 ns |    760.3081 ns |    711.1926 ns |      - |     152 B |
IgnoreList |  100000 |   570,505.43 ns |  1,213.7155 ns |  1,135.3102 ns |      - |     152 B |
IgnoreArray |  100000 |   418,304.10 ns |  1,100.5126 ns |  1,029.4201 ns |      - |     144 B |
Ignore | 1000000 | 4,670,008.96 ns | 12,812.1047 ns | 11,984.4496 ns |      - |     152 B |
IgnoreList | 1000000 | 5,713,345.04 ns | 12,848.2853 ns | 12,018.2929 ns |      - |     152 B |
IgnoreArray | 1000000 | 4,155,854.58 ns | 15,964.1574 ns | 14,932.8813 ns |      - |     144 B |

#### After

Method |       N |            Mean |          Error |         StdDev |  Gen 0 | Allocated |
------------ |-------- |----------------:|---------------:|---------------:|-------:|----------:|
Ignore |       1 |        45.47 ns |      0.6528 ns |      0.5096 ns | 0.0362 |     152 B |
IgnoreList |       1 |        44.05 ns |      0.6448 ns |      0.6031 ns | 0.0362 |     152 B |
IgnoreArray |       1 |        37.14 ns |      0.1212 ns |      0.1012 ns | 0.0343 |     144 B |
Ignore |      10 |        70.78 ns |      0.1940 ns |      0.1720 ns | 0.0361 |     152 B |
IgnoreList |      10 |        75.68 ns |      0.3982 ns |      0.3725 ns | 0.0362 |     152 B |
IgnoreArray |      10 |        58.53 ns |      0.2932 ns |      0.2742 ns | 0.0343 |     144 B |
Ignore |     100 |       334.72 ns |      0.6147 ns |      0.5750 ns | 0.0358 |     152 B |
IgnoreList |     100 |       384.79 ns |      1.4730 ns |      1.3058 ns | 0.0362 |     152 B |
IgnoreArray |     100 |       277.80 ns |      1.9549 ns |      1.7330 ns | 0.0343 |     144 B |
Ignore |    1000 |     2,918.57 ns |     14.7415 ns |     12.3098 ns | 0.0343 |     152 B |
IgnoreList |    1000 |     3,435.35 ns |     19.7293 ns |     17.4895 ns | 0.0343 |     152 B |
IgnoreArray |    1000 |     2,397.52 ns |     14.3753 ns |     12.7434 ns | 0.0343 |     144 B |
Ignore |   10000 |    28,638.57 ns |     70.1554 ns |     62.1909 ns | 0.0305 |     152 B |
IgnoreList |   10000 |    33,869.47 ns |    169.7860 ns |    158.8179 ns |      - |     152 B |
IgnoreArray |   10000 |    23,449.30 ns |     46.1209 ns |     40.8850 ns | 0.0305 |     144 B |
Ignore |  100000 |   285,544.17 ns |    399.3706 ns |    373.5715 ns |      - |     152 B |
IgnoreList |  100000 |   337,894.56 ns |  2,040.0016 ns |  1,592.6989 ns |      - |     152 B |
IgnoreArray |  100000 |   234,222.61 ns |  1,485.8049 ns |  1,389.8227 ns |      - |     144 B |
Ignore | 1000000 | 2,855,214.10 ns | 10,413.2421 ns |  9,231.0670 ns |      - |     152 B |
IgnoreList | 1000000 | 3,383,301.22 ns | 17,448.2561 ns | 16,321.1080 ns |      - |     152 B |
IgnoreArray | 1000000 | 2,340,943.35 ns | 11,849.7648 ns | 11,084.2762 ns |      - |     144 B |

**Comment by onovotny** [comment]
Interesting how much worse list is than array.

**Pull request merged**